lib/rubocop/cop/rails/date_and_time_calculation.rb

```diff
@@ -0,0 +1,725 @@
+      #
+      #   # good
+      #   date.yesterday?
+      #
+      #   # bad
+      #   Time.current - n.days
+      #   Time.zone.now - n.days
+      #
+      #   # good
+      #   n.days.ago
+      #
+      #   # bad
+      #   Time.current + n.days
+      #
+      #   # good
+      #   n.days.since
+      #
+      #   # bad
+      #   Date.current - 1
+      #   Date.current - 1.day
+      #
+      #   # good
+      #   Date.yesterday
+      #
+      #   # bad
+      #   Date.current + 1
+      #   Date.current + 1.day
+      #
+      #   # good
+      #   Date.tomorrow
+      #
+      #   # bad
+      #   time.after?(Time.current)
+      #   time > Time.current
+      #   Time.current < time
+      #   Time.current.before?(time)
+      #
+      #   # good
+      #   time.future?
+      #
+      #   # bad
+      #   time.before?(Time.current)
+      #   time < Time.current
+      #   Time.current > time
+      #   Time.current.after?(time)
+      #
+      #   # good
+      #   time.past?
+      class DateAndTimeCalculation < Base # rubocop:disable Metrics/ClassLength
+        extend AutoCorrector
+
+        CALCULATION_METHOD_NAMES = ::Set.new(%i[- +]).freeze
+
+        COMPARISON_METHOD_NAMES = ::Set.new(%i[< > after? before?]).freeze
+
+        DURATION_METHOD_NAMES = ::Set.new(
+          %i[
+            day
+            days
+            fortnight
+            fortnights
+            hour
+            hours
+            minute
+            minutes
+            month
+            months
+            second
+            seconds
+            week
+            weeks
+            year
+            years
+          ]
+        ).freeze
+
+        MSG = 'Prefer ActiveSupport date and time helper.'
+
+        RESTRICT_ON_SEND = [
+          *CALCULATION_METHOD_NAMES,
+          *COMPARISON_METHOD_NAMES,
+          :==,
+          :now,
+          :to_date,
+          :today,
+          :tomorrow,
+          :yesterday
+        ].freeze
+
+        # @param node [RuboCop::AST::SendNode]
+        # @return [void]
+        def on_send(node) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/MethodLength
+          case node.method_name
+          when :-
+            check_subtraction(node)
+          when :+
+            check_addition(node)
+          when :==
+            check_equality(node)
+          when :<, :before?
+            check_less_than(node)
+          when :>, :after?
+            check_greater_than(node)
+          when :now
+            check_now(node)
+          when :to_date
+            check_to_date(node)
+          when :today
+            check_today(node)
+          when :tomorrow
+            check_tomorrow(node)
+          when :yesterday
+            check_yesterday(node)
+          end
+        end
+
+        private
+
+        # @!method comparison_to_time_current?(node)
+        #   @param node [RuboCop::AST::Node]
+        #   @return [Boolean]
+        def_node_matcher :comparison_to_time_current?, <<~PATTERN
```

Comment:
Please move the `def_node_matcher` DSLs under constants. These are not published APIs and are not actually called externally. Mixing DSL and private methods is hard for maintainer to read.

Reply:
OK, I'll make that change 👌 

I think it is a common belief that methods should be kept `private` unless necessary, and that in cases where callbacks are realized in template method pattern like this, that callbacks should be defined in a more prominent position, but in this case you say we should do so even if it breaks these expectations, correct?

It is a bit strange that this point is being discussed at the time of the addition of this Cop, since some of the cops should have already implemented it this way. If there is such an implicit rule, it would be nice to have an internal affair cop or at least have some documentation for that.

Reply:
Reviewers don't always notice it, and they don't always point it out. Whether or not reviewers will comment will depend on the degree. Also, I don't want to retroactively change existing code because it is already in a sufficiently maintainable state.